### ISSUE TITLE
Enhancement: [no-unsafe-argument] <Improvements in type guard>

### ISSUE THREAD

**Issue by takemori-kondo** [typescript-eslint/typescript-eslint#6534]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unsafe-argument/

### Description

Suppress warnings in type guard.

The current no-unsafe-argument ignores type guarding.
It is useful if no-unsafe-argument understands type guards.

### Fail

```typescript
function foo(obj: object) {
  return obj;
}

let x: any = { a: 1, b: '2', c: { d: 3 } };

foo(x);
```

### Pass

```typescript
function foo(obj: object) {
  return obj;
}

let x: any = { a: 1, b: '2', c: { d: 3 } };
if (typeof x === 'object') {
  foo(x);
}
```

**Comment by armano2** [comment]
i don't think this is an issue, type of `x` before and after type guard in this case is any [try on playground](https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=GYVwdgxgLglg9mABMOcAUcBGArAXIrbAU2gEpEBvAWAChFEAnIqEBpQgbloF9baAbZogAe+AIZgAnogC8lRGPwBGADSJM+AOQAmTWoj4KiACb4AzIm6WuNWjGCI0USQAcicB8Nky5mwiShNcmo6ZFQ0YVIbbiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolp9oAc1gBbRC3RRE0aB2iRwYAL4h1QA)

![image](https://user-images.githubusercontent.com/625469/221434097-3311688a-ca92-4412-b682-5680c74be933.png)

if you want this type-guard to work you should to use `unknown` or any other type [try on playground](https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=GYVwdgxgLglg9mABMOcAUcBGArAXIrbAU2gEpEBvAWAChFEAnIqEBpQgbloF9baAbZogAe+cAGswcAO5IAvJUQBDfAEYANIkz4A5ACYdmiPgqIAJvgDMibja41aMYIjRQAngAcicZ8MRyAxB1CEigdcmo6ZFQ0YVJ7biA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolp9oAc1gBbRC3RRE0aB2iRwYAL4h1QA)

![image](https://user-images.githubusercontent.com/625469/221434079-1e716ed2-6833-44de-9da1-9679374fb094.png)

**Comment by bradzacher** [comment]
Yup - Armano is correct.
The source of truth for us is the TS types.

`any` is not refined away by TS so you're still passing `any` into the function.
